File: src/csharp/Intel/Iced.UnitTests/Intel/InstructionInfoTests/CodeValueTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO
using System.Collections.Generic;
using System.Text;
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.InstructionInfoTests {
	public sealed class CodeValueTests {
		[Fact]
		void Make_sure_all_Code_values_are_tested() {
			var tested = new bool[IcedConstants.CodeEnumCount];

			foreach (var info in GetTests())
				tested[(int)(Code)info[1]] = true;

			var sb = new StringBuilder();
			int missing = 0;
			var codeNames = ToEnumConverter.GetCodeNames();
			Assert.Equal(tested.Length, codeNames.Length);
			for (int i = 0; i < tested.Length; i++) {
				if (!tested[i] && !CodeUtils.IsIgnored(codeNames[i])) {
					sb.Append(codeNames[i] + " ");
					missing++;
				}
			}
			Assert.Equal("0 ins ", $"{missing} ins " + sb.ToString());
		}

		static IEnumerable<object[]> GetTests() {
			foreach (var info in InstructionInfoTest_16.Test16_InstructionInfo_Data)
				yield return info;
			foreach (var info in InstructionInfoTest_32.Test32_InstructionInfo_Data)
				yield return info;
			foreach (var info in InstructionInfoTest_64.Test64_InstructionInfo_Data)
				yield return info;
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/info/CodeValueTests.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

import java.util.ArrayList;

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.params.provider.*;

import com.github.icedland.iced.x86.CodeUtils;
import com.github.icedland.iced.x86.ToCode;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class CodeValueTests {
	@Test
	void make_sure_all_Code_values_are_tested() {
		boolean[] tested = new boolean[IcedConstants.CODE_ENUM_COUNT];

		for (Arguments args : getTests()) {
			InstructionInfoTestCase tc = (InstructionInfoTestCase)args.get()[0];
			tested[tc.code] = true;
		}

		StringBuilder sb = new StringBuilder();
		int missing = 0;
		String[] codeNames = ToCode.names();
		assertEquals(tested.length, codeNames.length);
		for (int i = 0; i < tested.length; i++) {
			if (!tested[i] && !CodeUtils.isIgnored(codeNames[i])) {
				sb.append(codeNames[i] + " ");
				missing++;
			}
		}
		assertEquals("0 ins ", String.format("%d ins %s", missing, sb.toString()));
	}

	private static Iterable<Arguments> getTests() {
		ArrayList<Arguments> result = new ArrayList<Arguments>();
		for (Arguments args : InstructionInfoTests.test16_InstructionInfo_Data())
			result.add(args);
		for (Arguments args : InstructionInfoTests.test32_InstructionInfo_Data())
			result.add(args);
		for (Arguments args : InstructionInfoTests.test64_InstructionInfo_Data())
			result.add(args);
		return result;
	}
}
